### PR TITLE
[GPU] Weightless caching

### PR DESCRIPTION
**Pull request by tkrupa-intel**

### CODE REVIEW

**Comment by p-durandin** [comment]
build_jenkins

**Comment by p-durandin** [comment]
build_jenkins

File: src/plugins/intel_gpu/src/plugin/program_builder.cpp [hunk 1]
```diff
@@ -304,6 +304,16 @@ void ProgramBuilder::add_primitive(const ov::Node& op, std::shared_ptr<cldnn::pr
     prim->origin_op_name = op.get_friendly_name();
     prim->origin_op_type_name = op.get_type_name();
 
+    if (auto data_prim = dynamic_cast<cldnn::data*>(prim.get())) {
+        auto rt_info = op.get_rt_info();
+        auto offset = rt_info.find("bin_offset");
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/src/plugin/program_builder.cpp:309]: I ensure it partially in two ways: 	
	1. I save bin_offset in Node's rt_info before the transformation pipeline is launched. This way if a Node is replaced by a new one during the trasformation pipeline, its values are cached in the traditional way. 
	2. During saving to cache I check if the byte size of the data has changed since the model was loaded - this mostly targets constants affected by the precision transformations.
		
I understand that this is a limited approach and there is a potential for more situations akin to the one with strided_slice in topologies I have not tested. I asked about it in the email thread you were just added to and received a response that it looks satisfactory. I you think it's not - do you have any suggestions on how to achieve it? I couldn't think of a 100% foolproof solution that does not involve direct comparison of the data (which would defeat the purpose).

File: src/plugins/intel_gpu/src/plugin/plugin.cpp [hunk 2]
```diff
@@ -162,6 +162,15 @@ Plugin::Plugin() {
     m_compiled_model_runtime_properties["OV_VERSION"] = ov_version.buildNumber;
 }
 
+void Plugin::set_cache_info(const std::shared_ptr<const ov::Model>& model, ExecutionConfig& config) const {
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/src/plugin/plugin.cpp:165]: Done.

File: src/plugins/intel_gpu/src/graph/program.cpp [hunk 3]
```diff
@@ -1723,8 +1723,40 @@ void program::cancel_compilation_context() {
 }
 
 void program::save(cldnn::BinaryOutputBuffer& ob) const {
+    std::string weights_path;
+    weights_path = _config.get_property(ov::intel_gpu::weights_path);
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/src/graph/program.cpp]: Done.

**Comment by p-durandin** [comment]
build_jenkins

File: src/plugins/intel_gpu/src/runtime/execution_config.cpp [hunk 4]
```diff
@@ -79,7 +79,8 @@ void ExecutionConfig::set_default() {
         std::make_tuple(ov::intel_gpu::allow_new_shape_infer, false),
         std::make_tuple(ov::intel_gpu::use_only_static_kernels_for_dynamic_shape, false),
         std::make_tuple(ov::intel_gpu::buffers_preallocation_ratio, 1.1f),
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/runtime/execution_config.cpp:81]: [random spot] please add some tests

**Comment by e-ddykim** [comment]
Did you check accuracy? When I tested this PR with the resnet-18 static model, the outputs are different between non-caching and weightless-caching runs.
Additionally, I temporarily commented out the below two lines for weightless cache blob loading:
https://github.com/openvinotoolkit/openvino/blob/7cf05641b6b1b249904c96e02ac07ee384219bb4/src/plugins/intel_gpu/src/plugin/plugin.cpp#L311-L312

**Comment by tkrupa-intel** [comment]
> Did you check accuracy? When I tested this PR with the resnet-18 static model, the outputs are different between non-caching and weightless-caching runs. Additionally, I temporarily commented out the below two lines for weightless cache blob loading:
> 
> https://github.com/openvinotoolkit/openvino/blob/7cf05641b6b1b249904c96e02ac07ee384219bb4/src/plugins/intel_gpu/src/plugin/plugin.cpp#L311-L312

Hi, thanks for letting me know about issues with this topology! I checked accuracy only for Stable Diffusion v1.5 and Llama-3-8b. I'm aware that there may be mismatches in other topologies (see discussion here: https://github.com/openvinotoolkit/openvino/pull/25731#discussion_r1756722874).

I'm aware that this check prevents correct import, I'll push the fix soon.

File: src/plugins/intel_gpu/src/runtime/execution_config.cpp (hunk 4 above)
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/src/runtime/execution_config.cpp:81]: What should they test specifically? There are no tests for traditional caching which I could use for reference. I wanted to compare constants' data directly in an original & imported models but I couldn't find a way to access that data using the standard API.

Comment (e-ddykim) [reply src/plugins/intel_gpu/src/runtime/execution_config.cpp:81]: You can find many test cases that end with '_cached' as below:
https://github.com/openvinotoolkit/openvino/blob/4758030e8e2c13383229c438aba945cdf86ad7a0/src/plugins/intel_gpu/tests/unit/test_cases/gemm_gpu_test.cpp#L3331-L3333
These tests are for the model caching feature.

File: src/plugins/intel_gpu/src/plugin/program_builder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/src/plugin/program_builder.cpp:309]: @vladimir-paramuzov I replaced the strided slice WA with a solution that universally prevents weightless caching for constants which are replaced during the transformation pipeline, even if the rt_info is explicitly copied from the old one to the new. Can you think of any other scenarios where this is not sufficient? Specifically, is it possible for constant's data to be altered without replacing a node with a new one during the transformation pipeline?

File: src/plugins/intel_gpu/src/runtime/execution_config.cpp (hunk 4 above)
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/src/runtime/execution_config.cpp:81]: Tests added.

**Comment by tkrupa-intel** [comment]
> Did you check accuracy? When I tested this PR with the resnet-18 static model, the outputs are different between non-caching and weightless-caching runs. Additionally, I temporarily commented out the below two lines for weightless cache blob loading:
> 
> https://github.com/openvinotoolkit/openvino/blob/7cf05641b6b1b249904c96e02ac07ee384219bb4/src/plugins/intel_gpu/src/plugin/plugin.cpp#L311-L312

Hi @e-ddykim, I get no mismatches with the current commit and a sample image. I tried to reproduce your issue with an old commit but got an exception throw instead of mismatches like you. Could you please recheck to make sure it's also fixed on your setup?

**Comment by p-durandin** [comment]
build_jenkins

File: src/core/src/rt_info.cpp [hunk 5]
```diff
@@ -74,7 +74,15 @@ ov::Any get_opset(const ov::Node::RTMap& rt_info) {
 
 void assign_runtime_info(const ov::Node::RTMap& from, ov::Node::RTMap& to) {
     auto opset = get_opset(to);
+
+    // Some items are designed to exist only in the original Node and disappear when copied to another one.
+    // This is done in case of items which are not applicable if their Node is replaced during the transformation
+    // pipeline.
+    std::vector<std::string> perishable_items = {"bin_offset", "original_size"};
```
Comment (vladimir-paramuzov) [inline src/core/src/rt_info.cpp:81]: I think you need to define a class derived from RuntimeAttribute instead of manipulating with string names:
```
class OPENVINO_API ConstantOriginalSize: public ov::RuntimeAttribute {
public:
    OPENVINO_RTTI("ConstantOriginalSize");
    ConstantOriginalSize() = default;
    bool is_copyable() const override {
        return false;
    }
};
```

and I think having is_copiable == false will resolve the issue mentioned above.

Comment (tkrupa-intel) [reply src/core/src/rt_info.cpp]: Yeah, that's definitely more elegant! Done.

**Comment by e-ddykim** [comment]
> > Did you check accuracy? When I tested this PR with the resnet-18 static model, the outputs are different between non-caching and weightless-caching runs. Additionally, I temporarily commented out the below two lines for weightless cache blob loading:
> > https://github.com/openvinotoolkit/openvino/blob/7cf05641b6b1b249904c96e02ac07ee384219bb4/src/plugins/intel_gpu/src/plugin/plugin.cpp#L311-L312
> 
> Hi @e-ddykim, I get no mismatches with the current commit and a sample image. I tried to reproduce your issue with an old commit but got an exception throw instead of mismatches like you, so we have different setups. Could you please recheck to make sure it's also fixed on your setup?

Now, I can get correct results from resnet-18 with this PR.

**Comment by p-durandin** [comment]
build_jenkins

File: src/core/include/openvino/op/util/weightless_caching_attributes.hpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+namespace ov {
+
+/**
+ * @brief ConstantOriginalSize class represents runtime info attribute that marks
+ * the original size of the constant in bytes. It is used in the weightless caching
+ * mechanism. It's not copyable in case the data was changed (the original node was
+ * replaced by a new one produced during the tranformation pipeline) - in that case
+ * weightless caching can't be used for that constant.
+ */
+class OPENVINO_API ConstantOriginalSize : public RuntimeAttribute {
```
Comment (praasz) [inline src/core/include/openvino/op/util/weightless_caching_attributes.hpp:19]: Move this attribute to dev API.
It looks like for now can OV internal attribute.

File: src/core/include/openvino/op/util/weightless_caching_attributes.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * @brief ConstantBinOffset class represents runtime info attribute that marks
+ * the binary offset of the constant's data in the weights file used by the
+ * weightless caching mechanism. It's not copyable in case the data
+ * was changed (the original node was replaced by a new one produced during the
+ * tranformation pipeline) - in that case weightless caching can't be used for
+ * that constant.
+ */
+class OPENVINO_API ConstantBinOffset : public RuntimeAttribute {
```
Comment (praasz) [inline src/core/include/openvino/op/util/weightless_caching_attributes.hpp:42]: Can we merge these two into single attribute as it looks one cannot exist without other. 
Rename it to `WeightlessCacheAttribute` which holds offset and size.

File: src/frontends/ir/src/frontend.cpp [hunk 8]
```diff
@@ -236,7 +237,12 @@ InputModel::Ptr FrontEnd::load_impl(const std::vector<ov::Any>& variants) const
         }
     }
 
-    return create_input_model();
+#if defined(OPENVINO_ENABLE_UNICODE_PATH_SUPPORT) && defined(_WIN32)
+    std::string weights_path_str = ov::util::wstring_to_string(weights_path);
+    return create_input_model(weights_path_str);
+#else
+    return create_input_model(weights_path);
+#endif
```
Comment (praasz) [inline src/frontends/ir/src/frontend.cpp:245]:
```suggestion
    return create_input_model(ov::util::path_to_string(weights_path));
```
?

File: src/frontends/ir/src/input_model.hpp [hunk 9]
```diff
@@ -22,7 +22,8 @@ class InputModel : public ov::frontend::InputModel {
 public:
     InputModel(std::istream& stream,
                const std::shared_ptr<ov::AlignedBuffer>& weights,
-               const std::unordered_map<ov::DiscreteTypeInfo, ov::BaseOpExtension::Ptr>& extensions);
+               const std::unordered_map<ov::DiscreteTypeInfo, ov::BaseOpExtension::Ptr>& extensions,
+               std::string weights_path = "");
```
Comment (praasz) [inline src/frontends/ir/src/input_model.hpp:26]:
```suggestion
               std::string weights_path = {});
```

File: src/frontends/ir/src/input_model.cpp [hunk 10]
```diff
@@ -205,13 +205,16 @@ class InputModel::InputModelIRImpl {
...
     pugi::xml_node m_root;
     pugi::xml_document m_xml_doc;
+    std::string m_weights_path;
 
 public:
     InputModelIRImpl(std::istream& stream,
                      const std::shared_ptr<ov::AlignedBuffer>& weights,
-                     const std::unordered_map<ov::DiscreteTypeInfo, ov::BaseOpExtension::Ptr>& extensions)
+                     const std::unordered_map<ov::DiscreteTypeInfo, ov::BaseOpExtension::Ptr>& extensions,
+                     std::string& weights_path)
```
Comment (praasz) [inline src/frontends/ir/src/input_model.cpp:214]:
```suggestion
                     std::string weights_path)
```

File: src/frontends/ir/src/input_model.cpp [hunk 11]
```diff
@@ -205,13 +205,16 @@ class InputModel::InputModelIRImpl {
...
         : m_weights(weights),
-          m_extensions(extensions) {
+          m_extensions(extensions),
+          m_weights_path(weights_path) {
```
Comment (praasz) [inline src/frontends/ir/src/input_model.cpp:217]:
```suggestion
          m_weights_path(std::move(weights_path)) {
```

File: src/frontends/ir/src/input_model.cpp [hunk 12]
```diff
@@ -227,8 +230,9 @@ class InputModel::InputModelIRImpl {
 
 InputModel::InputModel(std::istream& stream,
                        const std::shared_ptr<ov::AlignedBuffer>& weights,
-                       const std::unordered_map<ov::DiscreteTypeInfo, ov::BaseOpExtension::Ptr>& extensions) {
-    _impl = std::make_shared<InputModelIRImpl>(stream, weights, extensions);
+                       const std::unordered_map<ov::DiscreteTypeInfo, ov::BaseOpExtension::Ptr>& extensions,
+                       std::string weights_path) {
+    _impl = std::make_shared<InputModelIRImpl>(stream, weights, extensions, weights_path);
```
Comment (praasz) [inline src/frontends/ir/src/input_model.cpp:235]:
```suggestion
    _impl = std::make_shared<InputModelIRImpl>(stream, weights, extensions, std::move(weights_path));
```

File: src/core/include/openvino/op/util/weightless_caching_attributes.hpp (hunk 6 above)
Comment (tkrupa-intel) [reply src/core/include/openvino/op/util/weightless_caching_attributes.hpp:19]: Do I move to dev API just by deleting OPENVINO_API or replacing it with something else? I didn't find any preprocessor define which has DEV.*API in the name and if I delete OPENVINO_API I get linker errors while building program_builder.cpp.

Comment (praasz) [reply src/core/include/openvino/op/util/weightless_caching_attributes.hpp:19]: Move header to `src/core/dev_api/openvino/core/rt_info/*.hpp`, keep `OPENVINO_API` macro.

File: src/frontends/ir/src/ir_deserializer.cpp [hunk 13]
```diff
@@ -944,6 +945,12 @@ std::shared_ptr<ov::Node> ov::XmlDeserializer::create_node(const std::vector<ov:
         if (aw_data) {
             rtInfo["alt_width"] = aw_data.value();
         }
+        if (auto constant = std::dynamic_pointer_cast<ov::op::v0::Constant>(ovNode)) {
+            rtInfo[ov::ConstantBinOffset::get_type_info_static()] =
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/frontends/ir/src/ir_deserializer.cpp:949]: I'm not sure I understand, could you rephrase this comment?

If I remove the if statement and set rtInfo unconditionally, I get:
`Exception from src/inference/src/cpp/core.cpp:90:
Check 'false' failed at src/frontends/common/src/frontend.cpp:53:
Converting input model
node <data> is missing mandatory attribute: size at offset 137
`

[the remark this answers is not in the document]
Comment (praasz) [reply src/frontends/ir/src/ir_deserializer.cpp:949]: Sure,
I mean to check if attribute exists in xml then add it to node as `rt_info`. 
This attribute is not mandatory and current xml will not have it. If not existing just skip add rt_info attribute.

File: src/core/tests/pass/serialization/deterministicity.cpp [hunk 14]
```diff
@@ -193,6 +193,12 @@ TEST_P(SerializationDeterministicityInputOutputTest, FromOvModel) {
     auto& expected1 = modelRef;
     ov::pass::Serialize(m_out_xml_path_1, m_out_bin_path_1, irVersion).run_on_model(modelRef);
     auto expected2 = ov::test::readModel(m_out_xml_path_1, m_out_bin_path_1);
+
+    // We need to check and erase this entry because it's expected to be different in these two scenarios.
+    std::string path_key = "weights_path";
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/core/tests/pass/serialization/deterministicity.cpp:198]: The problem is that the function reading the model is plugin-agnostic - it will set the weights_path regardless of the plugin(s) which will be used with the model later. Do you suggest that the function reading the model should set the weights_path conditionally?

**Review by t-jankowski — CHANGES_REQUESTED** [review]

File: src/core/include/openvino/op/util/weightless_caching_attributes.hpp [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+public:
+    OPENVINO_RTTI("ConstantOriginalSize");
+
+    ConstantOriginalSize() = default;
+
+    ConstantOriginalSize(size_t original_size) : original_size(original_size) {}
+
+    bool is_copyable() const override {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/core/include/openvino/op/util/weightless_caching_attributes.hpp]: Done.

File: src/core/include/openvino/op/util/weightless_caching_attributes.hpp [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+//
+
+#pragma once
+
+#include "openvino/core/core_visibility.hpp"
+#include "openvino/core/runtime_attribute.hpp"
+
+namespace ov {
+
+/**
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/core/include/openvino/op/util/weightless_caching_attributes.hpp]: Done.

File: src/core/include/openvino/op/util/weightless_caching_attributes.hpp (hunk 6 above)
Comment (tkrupa-intel) [reply src/core/include/openvino/op/util/weightless_caching_attributes.hpp]: Done.

File: src/core/tests/pass/serialization/rt_info_serialization.cpp [hunk 17]
```diff
@@ -257,6 +257,7 @@ TEST(OvSerializationTests, SerializeRawMeta) {
 		<custom_rt_info2>
 			<item0 value="testvalue2" />
 		</custom_rt_info2>
+		<weights_path value="" />
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/core/tests/pass/serialization/rt_info_serialization.cpp]: Removed weights_path from model's rt_info altogether. Now it's expected to be provided via config to calls to API functions such as compile_model() and import_model().

File: src/core/tests/pass/serialization/deterministicity.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/core/tests/pass/serialization/deterministicity.cpp]: Removed - it's no longer serialized.

File: src/frontends/ir/src/ir_deserializer.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/frontends/ir/src/ir_deserializer.cpp]: Done.

File: src/frontends/ir/src/input_model.cpp [hunk 18]
```diff
@@ -244,6 +248,7 @@ std::shared_ptr<ov::Model> InputModel::InputModelIRImpl::convert() {
     std::shared_ptr<ov::Model> model;
     visitor.on_attribute("net", model);
     model->get_rt_info()["version"] = int64_t(version);
+    model->get_rt_info()["weights_path"] = m_weights_path;
```
[the remark this answers is not in the document]
Comment (tkrupa-intel) [reply src/frontends/ir/src/input_model.cpp]: Done.

**Comment by p-durandin** [comment]
build_jenkins

File: src/core/dev_api/openvino/core/rt_info/weightless_caching_attributes.hpp [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    WeightlessCacheAttribute() = default;
+
+    WeightlessCacheAttribute(size_t original_size, size_t bin_offset)
+        : original_size(original_size),
+          bin_offset(bin_offset) {}
+
+    bool is_copyable() const override;
+
+    size_t original_size;
+    size_t bin_offset;
```
Comment (t-jankowski) [inline src/core/dev_api/openvino/core/rt_info/weightless_caching_attributes.hpp:35]: Set default values for members or consider removing default constructor.

File: src/core/tests/pass/serialization/deterministicity.cpp
Comment (praasz) [inline src/core/tests/pass/serialization/deterministicity.cpp:1]: Can we have test where to parse xml with new attributes to check if deserialization works correctly?
Add test for serialization where ov model is exported to IR (xml)?

File: src/core/dev_api/openvino/core/rt_info/weightless_caching_attributes.hpp (hunk 19 above)
Comment (tkrupa-intel) [reply src/core/dev_api/openvino/core/rt_info/weightless_caching_attributes.hpp]: Deleted default constructor.

File: src/core/tests/pass/serialization/deterministicity.cpp
Comment (tkrupa-intel) [reply src/core/tests/pass/serialization/deterministicity.cpp:1]: My understanding is that only model's rt_info is serialized while nodes' rt_info is not serialized. Is my understanding correct? If so, the new attributes are not a part of the serialization process.

**Comment by p-durandin** [comment]
build_jenkins

File: src/core/tests/pass/serialization/deterministicity.cpp
Comment (praasz) [reply src/core/tests/pass/serialization/deterministicity.cpp:1]: [Here](https://github.com/openvinotoolkit/openvino/blob/485e5802ec8bf402dc27db1e60f782cede28568c/src/core/src/pass/serialize.cpp#L1010) node's rt info is serialized (not all attributes can be it requires some additional work).

If this new attribute will be set by e.g. GPU plugin then after export to xml the new attribute will be lost and when read xml again there will be no information set it again.

If required it can be done as separate PR.

**Review by praasz — APPROVED** [review]

**Review by t-jankowski — APPROVED** [review]
Lgtm regarding Core and IR FE

File: src/plugins/intel_gpu/include/intel_gpu/runtime/internal_properties.hpp [hunk 20]
```diff
@@ -56,6 +56,7 @@ static constexpr Property<size_t, PropertyMutability::RW> max_dynamic_batch{"DYN
 static constexpr Property<bool, PropertyMutability::RW> nv12_two_inputs{"GPU_NV12_TWO_INPUTS"};
 static constexpr Property<float, PropertyMutability::RW> buffers_preallocation_ratio{"GPU_BUFFERS_PREALLOCATION_RATIO"};
 static constexpr Property<size_t, PropertyMutability::RW> max_kernels_per_batch{"GPU_MAX_KERNELS_PER_BATCH"};
+static constexpr Property<std::string, PropertyMutability::RW> weights_path{"GPU_WEIGHTS_PATH"};
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/include/intel_gpu/runtime/internal_properties.hpp:59]: Is it really GPU specific property? Can it theoretically be reused for other plugins weightless cache impl?

**Comment by p-durandin** [comment]
build_jenkins

File: src/inference/include/openvino/runtime/properties.hpp [hunk 21]
```diff
@@ -1345,4 +1345,10 @@ static constexpr Property<Affinity> affinity{"AFFINITY"};
  */
 static constexpr Property<std::vector<std::string>, PropertyMutability::RO> execution_devices{"EXECUTION_DEVICES"};
 
+/**
+ * @brief Path to the file with model's weights.
+ *
+ * @note This property is used for weightless caching. Only used when ov::CacheMode Property is set to "OPTIMIZE_SIZE".
+ */
+static constexpr Property<std::string, PropertyMutability::RW> weights_path{"WEIGHTS_PATH"};
```
Comment (akuporos) [inline src/inference/include/openvino/runtime/properties.hpp:1353]: I guess it should be added to python api files as well: [#1](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/pyopenvino/core/properties/properties.cpp) [#2](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/openvino/runtime/properties/__init__.py) [#3](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/tests/test_runtime/test_properties.py)

File: src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+void CheckWeightlessCacheAccuracy::TearDown() {
+    std::remove(xml_path.c_str());
+    std::remove(bin_path.c_str());
+    std::remove(cache_path.c_str());
+}
+
+void CheckWeightlessCacheAccuracy::run() {
+    ov::AnyMap config = {{"CACHE_MODE", ov::CacheMode::OPTIMIZE_SIZE}, {"WEIGHTS_PATH", bin_path}};
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:63]: `{ ov::cache_mode(ov::CacheMode::OPTIMIZE_SIZE), ov::weights_path(bin_path) }`

File: src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+    auto core = ov::test::utils::PluginCache::get().core();
+    ov::pass::Serialize(xml_path, bin_path).run_on_model(model);
+
+    ov::CompiledModel compiled_model;
+    OV_ASSERT_NO_THROW(compiled_model = core->compile_model(xml_path, ov::test::utils::DEVICE_GPU, config));
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:68]: I wonder if Core can automatically set weights path when model is loaded from IR? Would be nice if users won't need to set this property manually all the time

Comment (tkrupa-intel) [reply src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:68]: It was like that in an earlier version but I changed it after it was requested that it should only exist in runtime: https://github.com/openvinotoolkit/openvino/pull/25731#discussion_r1799301914.

File: src/inference/include/openvino/runtime/properties.hpp (hunk 21 above)
Comment (tkrupa-intel) [reply src/inference/include/openvino/runtime/properties.hpp:1353]: Done.

File: src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp (hunk 22 above)
Comment (tkrupa-intel) [reply src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp]: Done.

**Review by akuporos — APPROVED** [review]

**Comment by p-durandin** [comment]
build_jenkins

File: src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp (hunk 23 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:68]: My suggestion is a little bit different. I agree that weighs path shouldn't be a part of serialized IR. But we may do something like this
```
ov::Model model = core.read_model(...); // it saves weights path as non-serializable attribute in ov::Model
ov::CompiledModel = core.compile_model(model); // internally core converts this attribute of ov::Model to ov::weights_path property and pass it to the plugin
```

In that case users can avoid setting weights path explicitly and it will be mainly needed if they call `import_model` explicitly

Comment (tkrupa-intel) [reply src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:68]: Sure, that sounds good. What about saving it in the cache file? I would assume it shouldn't be exported to cache for the same reason it shouldn't be serialized in the IR?

Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:68]: Yes, I think it shouldn't be exported to the cache too.

Comment (tkrupa-intel) [reply src/plugins/intel_gpu/tests/functional/behavior/model_cache.cpp:68]: > ov::Model model = core.read_model(...); // it saves weights path as non-serializable attribute in ov::Model

I contacted the core/serialization team and received information that currently all model attributes are serialized and it's impossible to have a non-serializable one without modifying the serialization mechanism itself. I feel that such change is outside of scope of this PR. What are your thoughts on this?

**Review by vladimir-paramuzov — APPROVED** [review]

File: src/inference/include/openvino/runtime/properties.hpp (hunk 21 above)
Comment (ilya-lavrenov) [inline src/inference/include/openvino/runtime/properties.hpp:1353]: why is it public property? Is external user supposed to set it?

Can it be moved to https://github.com/openvinotoolkit/openvino/blob/master/src/inference/dev_api/openvino/runtime/internal_properties.hpp ?

Comment (tkrupa-intel) [reply src/inference/include/openvino/runtime/properties.hpp:1353]: Yes, it's not saved in the cache, so the external user has to provide it like here: https://github.com/openvinotoolkit/openvino/pull/25731/files#diff-305779c23a13e2f28656783813ddbdbd8b962d20e3ec05683e52ebe89d0df259R63.

Comment (ilya-lavrenov) [reply src/inference/include/openvino/runtime/properties.hpp:1353]: As @vladimir-paramuzov and @slyalin wrote, we need to patch frontends to add this attribute to `ov::Model`, so users don't need to deal with this attribute.

Comment (tkrupa-intel) [reply src/inference/include/openvino/runtime/properties.hpp:1353]: Yes, I'm working on this right now. This will eliminate the need to provide it to compile_model() by user. However, the user will still have to provide it to import_model() once they intend to load the model from cache.

**Review by p-durandin — APPROVED** [review]

**Comment by p-durandin** [comment]
build_jenkins

**Comment by p-durandin** [comment]
build_jenkins
